### DISCUSSION TITLE
Current emscripten version/setup?

### DISCUSSION THREAD

**Discussion by craigsapp** [rism-digital/verovio#3860] [Q&A] [answered]
What is the current recommended setup for compiling veorvio with emscripten?

I have upgraded to MacOS Sequoia 15.1, with clang++ v16.0.0 (clang++ v15 needed for full C++20 support), and emsdk/emcc 3.1.45.   Mostly upgrade was to try to compile C++20 code with emscripten which my previous version of emsdk/emcc was not liking.

I tried the latest version of emsdk and it did not work, so I chose emsdk 3.1.45 based on this thread a year ago recommending emsdk 3.1.45 (as opposed to 3.1.47):

https://github.com/rism-digital/verovio/issues/3533#issuecomment-1758960456

---------------------

I get the warning `invalid item in EXPORTED_RUNTIME_METHODS: HEAPU8` which may be significant and some stuff about node (I have node v23.2.0 installed, but emsdk seems to be using its own node version 20.18.0).

```
Building makefile...
configure: cmake ../cmake -DBUILD_AS_WASM=ON "-DCMAKE_CXX_FLAGS=-O3 -DNDEBUG -std=c++20 -s STRICT=1" -DCMAKE_TOOLCHAIN_FILE=/Users/craig/git-cloud/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/Users/craig/git-cloud/emsdk/node/20.18.0_64bit/bin/node
-- ***** Building Verovio as WASM BC *****
-- Configuring done (3.4s)
-- Generating done (0.1s)

[compiling src files without problems]

Linking...
cache:INFO: generating system asset: symbol_lists/5d11dd39ac41f9ef8682278725647c9cf5091319.json... (this will be cached in "/Users/craig/git-cloud/emsdk/upstream/emscripten/cache/symbol_lists/5d11dd39ac41f9ef8682278725647c9cf5091319.json" for subsequent builds)
cache:INFO:  - ok
warning: invalid item in EXPORTED_RUNTIME_METHODS: HEAPU8
em++: warning: warnings in JS library compilation [-Wjs-compiler]
 Done.
(node:28575) ExperimentalWarning: CommonJS module /opt/homebrew/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/homebrew/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
```

And here is the resulting problem when running in a browser:

<img width="818" alt="Screenshot 2024-11-13 at 20 38 50" src="https://github.com/user-attachments/assets/e308be59-3429-4123-b167-1463b1d09f3c">

**Comment by lpugin** [comment]
I am currently using `3.1.65`, and so is the cache used in [GH actions](https://github.com/rism-digital/verovio/actions/runs/11779747631/job/32808862730#step:5:29). Could you try that one?

**Reply by lpugin** [reply]
I just gave a try with the latest `3.1.71` and it builds properly.

**Comment by craigsapp** [comment]
Thanks: I am not seeing this warning anymore with version 3.1.71: `invalid item in EXPORTED_RUNTIME_METHODS: HEAPU8`

However, I still get the error:

<img width="818" alt="Screenshot 2024-11-13 at 20 38 50" src="https://github.com/user-attachments/assets/6c9a7df5-589f-4399-90ff-674d8691b45b">

Have you done something with the humdrum buffer (since September 27)?  I have not found anything significant in the `Toolkit` class or in `emscripten/buildToolkit` yet.

----------------------------------

This test HTML works:

```html
<html>
<head>
<script src="https://www.verovio.org/javascript/latest/verovio-toolkit-hum.js" defer></script>

<script>
document.addEventListener("DOMContentLoaded", (event) => {
   verovio.module.onRuntimeInitialized = () => {
      let tk = new verovio.toolkit();
      console.log("Verovio has loaded!");
      let svg = tk.renderData("**kern\n1c\n*-\n", {});
      document.querySelector("#notation").innerHTML = svg;
   }
});
</script>

</head>
<body>
<h1>Hello Verovio!</h1>
<div id="notation"></div>
</body>
</html>
```

producing the webpage:

<img width="774" alt="Screenshot 2024-11-14 at 05 54 43" src="https://github.com/user-attachments/assets/e57d51ca-d544-4916-9a48-466706e92712">

But when I call the `tk.getHumdrum()` function:

```html
<html>
<head>
<script src="https://www.verovio.org/javascript/latest/verovio-toolkit-hum.js" defer></script>

<script>
document.addEventListener("DOMContentLoaded", (event) => {
   verovio.module.onRuntimeInitialized = () => {
      let tk = new verovio.toolkit();
      console.log("Verovio has loaded!");
      let svg = tk.renderData("**kern\n1c\n*-\n", {});
      document.querySelector("#notation").innerHTML = svg;
      let humdrum = tk.getHumdrum();
      document.querySelector("#humdrum").innerHTML = humdrum;
   }
});
</script>

</head>
<body>
<h1>Hello Verovio!</h1>
<div id="notation"></div>
<div id="humdrum"></div>
</body>
</html>
```

I get the same error:

<img width="1344" alt="Screenshot 2024-11-14 at 05 58 33" src="https://github.com/user-attachments/assets/af15d60e-acd3-4b1d-ba37-39b9804c1e98">

**Reply by lpugin** [reply, marked as answer]
I think [this](https://github.com/rism-digital/verovio/commit/49316cab2717e60e633b0b131fb18af9ce01c1a2) should fix it. I am not sure how it ever worked before.

**Reply by craigsapp** [reply]
Thanks!  That fixed the problem.  I was about to go look for the js interface functions and was noticing they were no longer in `emscripten` (now moved to `emscripten/npm/src`).   Maybe moving them into npm or some update to emcc revealed the problem (which I would not have noticed myself :-).

**Reply by craigsapp** [reply]
I will start to look into issue https://github.com/rism-digital/verovio/issues/3827 now.
